binding/c: update python scripts

hzhou opened the pull request:
## Pull Request Description

Split from PR #4941 

This PR updates the python binding scripts as a per-requisite for the rest of the API generation. 

The actual impact of this PR is when we enable the generation of rest of API functions. We can always make necessary patches then.

Added --

Ability and option to directly load `apis.json` as from the standard repository build process. If we import the `apis.json` to `src/binding/` folder, it will load apis.json instead of `mpi_standards_api.txt`.

Use `re.IGNORECASE` instead of `re.I`

Better temporary variable names.

Added `.docnotes` directive for additional docnotes

Added `.errorcodes` directives for additional error codes.

## Expected Impact

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [ ] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

hzhou commented:
@shintaro-iwasaki I think you already reviewed most of the commits in this PR. Could review it again or approve it while I make sure the tests are good?

shintaro-iwasaki reviewed (approved):
Except for "binding/c: update python scripts" which is too large to review, this PR looks okay to me.

hzhou commented:
test:mpich/ch4/most

raffenet commented on maint/gen_binding_c.py:
```diff
@@ -5,28 +5,47 @@
 from local_python import MPI_API_Global as G
 from local_python.mpi_api import load_mpi_api
-from local_python.binding_c import dump_mpi_c, dump_Makefile_mk
+from local_python.binding_c import *
```
it is not possible for me to do any meaningful review of this commit. there are 1500 lines of change. i don't know what to do with it besides look at the output.

hzhou replied:
I know. Unless there is something you particularly notice, such as styles or obvious strangeness, I think we can merge them. They should not have impact to the current code other than the pt2pt chapter that already got merged. We'll review the output in the following chap-by-chap PRs and if there is errors, we can always add patches then.

raffenet commented on maint/gen_binding_c.py:
```diff
@@ -5,28 +5,47 @@
+    func_list = [f for f in G.FUNCS.values() if 'dir' in f]
+    func_list.sort(key = lambda f: f['dir'])
+    for func in func_list:
+        if RE.search(r'not_implemented', func['attrs']):
```
IMO, regex is overkill for simple substring matches. This can be done with `in` or `.find()`, with the benefit that it should perform better.

hzhou replied:
I can change that.

Regex is fast enough, so I wouldn't worry about performance optimization unless there is evidence and justification. One of the drawback is regex tend to become too complex. That can be fixed by restructure the code design to keep the regex simple.

The benefit of regex vs direct string comparison or `in` or `find` is flexibility. For example, it can easily accommodate extra spaces, multiple spelling variations, flexible cases, etc. without significantly change the code appearance or structure. The alternative of using a mix of `==`, `in`, `find`, and regex, imo, can introduce unnecessary visual code complexity.

So I hope I convinced you to overlook this detail, but if you insist, I'll change it (and potentially many other places).

raffenet commented on maint/local_python/binding_c.py:
```diff
@@ -133,84 +200,173 @@ def process_func_parameters(func):
             t = ''
             for j in range(group_count):
-                a = func['params'][i + j]['name']
+                pj = func['params'][i + j]
```
what is `pj`?

hzhou replied:
A temporary variable name for `p` since `p` is already used.  Will `p_j` reads better?

hzhou replied:
changed into `temp_p`

raffenet commented on maint/local_python/binding_c.py:
```diff
@@ -397,70 +638,212 @@ def dump_function(func, map_name):
+            if kind in G.handle_NULLs:
+                G.out.append("*%s = %s;" % (name, G.handle_NULLs[kind]))
+    elif RE.match(r'mpi_type_', func['name'], re.I):
+        p = func['params'][-1]
```
Single letter variable names are really hard to understand. We should fix all the instances in these scripts.

hzhou replied:
There are only a couple --
* `G` for globals. It is more of a marker since Python has no facility to distinguish global or local variables and it has no scopes. 
*  `p` for parameter, `param` better?
*  `a` used for parsing arguments and lists, it is scoped within a few lines. I often find it more confusing for very local temporary variables to have longer names. The latter the reader will tempted to search for meanings elsewhere while there isn't any. It is similar to `i`, and `j`.

raffenet commented on maint/local_python/binding_c.py:
```diff
@@ -397,70 +638,212 @@ def dump_function(func, map_name):
+            G.out.append("%s *%s_ptr = NULL;" % (mpir_type, name))
+            if kind in G.handle_NULLs:
+                G.out.append("*%s = %s;" % (name, G.handle_NULLs[kind]))
+    elif RE.match(r'mpi_type_', func['name'], re.I):
```
We should use the long form `re.IGNORECASE` to make intent clearer.

hzhou replied:
Okay

hzhou replied:
mark as fixed (once pushed)

raffenet commented on src/include/mpir_impl_tmp.h:
```diff
@@ -3,8 +3,8 @@
 
-#ifndef MPIR_IMPL_H_INCLUDED
-#define MPIR_IMPL_H_INCLUDED
+#ifndef MPIR_IMPL_TMP_H_INCLUDED
```
I feel we can separately review the generation of `mpir_impl.h` after the MPI/PMPI work is merged. Or is there some reason this needs to be done together?

hzhou replied:
If we don't generate `mpir_impl.h`, then I have to manually declare the prototypes, as I did with `pt2pt` and during refactoring of `attr` and `errorhan`. That is the reason `mpir_impl.h` was created in the first place. We have to manually check which function needs the prototypes since many but not all in every chapter has slightly changed functions names (mostly with or without `_impl` suffix).  This is a very tedious work that will get deleted in the end. So I am hoping that we can avoid it.

raffenet commented on src/mpi/errhan/errutil.c:
```diff
@@ -1656,6 +1693,10 @@ static int vsnprintf_mpi(char *str, size_t maxlen, const char *fmt_orig, va_list
                     MPL_snprintf(str, maxlen, "errh=0x%x", E);
                 }
                 break;
+            case (int) 'K':
```
Is this needed? Or just an improvement to the error output? I can't see why returning the predefined integer values is wrong.

hzhou replied:
It is for consistency. I don't see why the KEYVAL handle object is any different from other handle objects. The integer value wouldn't make any sense to the user -- at least we should print it as hexadecimals.

raffenet commented on maint/mpi_standard_api.txt:
```diff
@@ -150,15 +150,15 @@ MPI_Alltoallw_init:
     request: REQUEST, direction=out, [communication request]
 MPI_Attr_delete:
-    comm: COMMUNICATOR, direction=lis:inout, param:in, [communicator to which attribute is attached]
+    comm: COMMUNICATOR, direction=in, [communicator to which attribute is attached]
```
has the script used to scrape the standard data been committed to the repo?

hzhou replied:
No. 

The mpi standards -- at least the API functions -- are not expected to change often. I think it is simpler to directly maintain `mpi_standard_api.txt` directly (even manually), rather than having to maintain one more script and add a dependency on the private mpi-standards repository.

raffenet replied:
I don't think this is a good idea. There is a hidden step to this that only you currently understand.

hzhou replied:
If I say I manually transcribed the api interface into this config file, would that make any difference?

The script I use is in Perl and it is for one-time purpose that I don't intend to maintain (other than hack). This config file will always need be directly maintained anyway since we want to review explicit changes from this point on. If we get initial transcription correct, I expect it is not a big task in the future to manually update this file.

It is one of the intentions to decouple this config file from the python-framework used in mpi-standards. That python layer are subject to change with no apparent reasons from outside. For example, they may change the way how the categorize parameters, variable names, particular attributes syntaxes, etc.. I don't think we really want to follow those changes. Any changes to `mpi_standard_api.txt` in the future should be treated as an update from standards from user point of view.

hzhou replied:
Besides, if we want a *correct* scraper, we'll need the scraper to be able to parse full Python syntax. That is way too complex and unnecessary. But without the *correctness*, the script is considered hacky and thus I am unwilling to maintain.

raffenet replied:
> It is one of the intentions to decouple this config file from the python-framework used in mpi-standards.

I disagree with this intent. The source data in the MPI standard is json. Python includes a json parser in the standard library. I don't think this is a big ask.

hzhou replied:
It is not json. It is full python code:
```
\begin{mpi-binding}
    function_name("MPI_Send")

    parameter(
        "buf", "BUFFER", desc="initial address of send buffer", constant=True
    )
    parameter(
        "count",
        "XFER_NUM_ELEM_NNI_SMALL",
        desc="number of elements in send buffer",
    )
    parameter(
        "datatype", "DATATYPE", desc="datatype of each send buffer element"
    )
    parameter("dest", "RANK", desc="rank of destination")
    parameter("tag", "TAG", desc="message tag")
    parameter("comm", "COMMUNICATOR")
\end{mpi-binding}
```

To do it "correctly", not only we need parse python syntax, we also need import all their python stack. Probably that the only way is to directly import their code, then we are having a hard dependency on it.

raffenet replied:
If you build the document, there is an `apis.json` file generated.

hzhou replied:
Also their python code is focused on generating 'LaTeX', so importing it still doesn't reduce the complexity much on our part.

raffenet replied:
I'm not talking about importing their Python code. Please don't misunderstand me.

hzhou replied:
I didn't realize the generated `apis.json`. We can directly use the json file -- i.e. directly add `apis.json` to our repository. Is that more preferable. The apiis.json doesn't have language bindings (parameter types), so that still need be separately scraped.

raffenet replied:
It might be worth engaging the standard maintainers to see if parameter types might be something they can add to the json, or if some other solution exists.

I don't want to block others, so I think we can move forward with this. The API source information is a big piece of technical debt that I think we should address in the long-term.

hzhou commented:
@raffenet I added two additional commits. The last one loads `apis.json` directly. `apis.json` is 56,254 lines long 😱 (vs. `mpi_standard_api.txt` 3,432 lines) 😄 Anyway, could check if the new version is preferrable?

hzhou commented:
@raffenet I removed the `apis.json` but retained the ability to load `apis.json`, so we always have a fallback option and it is good if we want to verify the correctness of `mpi_standards_api.txt`.

I addressed your comment on `re.IGNORECASE` and better temporary variable name.

I added `.docnotes` and `.errorcodes` directives to reflect the review from @shintaro-iwasaki. 

Do you see any other issues I need to address for this PR?

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

raffenet reviewed (approved):
We had a lot of offline discussion. Short-term, I think this is fine to merge and unblock everyone else.

Long-term, I hope we can better leverage what the MPI standard document provides and avoid the custom data format and parser API information. I have reached out to the people maintaining this info in the standard, and `apis.json` is currently the best source of API data. However, more work is planned to better facilitate code generation and they are looking for input from folks like us. I think it would be mutually beneficial if we can contribute to their efforts.

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
@raffenet I added two more addition trivial commits to this PR. If there is no further comments, I will merge after tests clear. Otherwise, let me know.
